docs(rfc): add api access modifiers style-guide RFC for v9 (#23577)
microsoft/fluentui: rfcs/shared/build-system/package-public-api.md

## Problem statement

We expose all APIs of our packages as a PUBLIC from barrel files (`src/index.ts`) to consumers, which introduces problems as we can expose apis that were supposed to be used only internally within our repo or only be available in particular pre-release stage (ALPHA/BETA versions).

## Detailed Design or Proposal

We wanna provide style guide/common approach how to annotate package APIs with proper ACCESS modifiers which will be processed/verified by tooling automation.

Note that our approach will be applied only on **Type Declaration emit level** as we consider TypeScript surface as only source of truth for our package APIs.

> This means that if users use only vanilla JavaScript nothing will stop them from using those API's - as we don't support that kind of usage users take the risk that things may/will break for them unfortunately and also they will not get any support from our side.

### Proposed access modifiers:

> #### `@internal`

Reviewer: We could have @public be the modifier and have the tooling make all things @internal by default.  This would force us to be explicit in what is exposed.  If we forget the attribute, the method or type can't get published.

Reviewer 2: I actually agree with this 👍 how easy it is to implement might be a differnt story

Author: interesting idea indeed @GeoffCoxMSFT ! that would work only for stable API's though, which is not our use-case as we have various stages of API/packages that are being shipped to users.

> how easy it is to implement might be a differnt story

haha YES

It indicates that an API item is meant only for usage by other NPM packages from the same repo. Third parties _should never use “internal” APIs._

- tooling will trim the declaration from a public API surface.

> - tolling will trim the declaration from `/unstable` API surface.

Reviewer 2: ```suggestion
- tooling will trim the declaration from `/unstable` API surface.
```

**Example:**

```ts
// @filename package-a/src/index.ts

/**
 *
 * @internal
 */
export function doSomething(): void {}

export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/index.d.ts

export declare function hello(): void;
```

> #### `@alpha`

Reviewer 2: I'm not quite sure how useful the prerelease stage modifiers will be since it's harder for partners to be aware of these tags, not all IDEs will render tsdoc the same way. For the scenarios that the prerelease modifiers cover, IMO `/unstable` entrypoint + `_unstable` suffix would be more impactful

Reviewer 3: > additional context: if needed we can allow @beta to be shipped from stable and only @alpha from unstable but I'd leave that to team decisions ( IMO we should not do that ).

As `@beta` packages could have breaking changes IMO we should not ship them under stable.

To check my understanding: we have following chain for packages `alpha` -> `beta` -> `rc`. If we follow the same for APIs in comments, we should have also `@rc` comment, right? 👾

Author: > does this mean that @fluentui/new-component will only be usable from @fluentui/new-component/unstable while the component is in alpha and beta stages ?

yes that's the idea (from TS api POV _

It indicates that an API item is eventually intended to be public, but currently is in an early stage of development

- tooling will trim the declaration from a public API surface.
- can be exposed only from `/unstable` API surface
- tolling will NOT TRIM the declaration from `/unstable` API surface.

**Example:**

- unstable api

```ts
// @filename package-a/src/unstable/index.ts

/**
 *
 * @alpha
 */
export function doSomething(): void {}

export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/unstable.d.ts
export function doSomething(): void;
export declare function hello(): void;
```

- stable api

```ts
// @filename package-a/src/index.ts

// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
/**
 *
 * @alpha
 */
export function doSomething(): void {}
/**
 *
 * @public
 */
export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/index.d.ts
export declare function hello(): void;
```

> #### `@beta`

Reviewer 3: I don't see any difference between in `@alpha` & `@beta` except semantics, is correct? As we export them as under unstable is there a sense to use `@unstable` instead?

IMO it will reduce maintenance cost as with current proposal we should ensure that `@alpha` will be changed to `@beta` when a package will be promoted. And we don't have any automation for it...

Author: > I don't see any difference between in @alpha & @beta except semantics, is correct 

correct. I added more context to one of Lings comments
↓
![image](https://user-images.githubusercontent.com/1223799/174262747-0a5d0cd0-e595-4346-bcf2-209e58b0e4ce.png)

> As we export them as under unstable is there a sense to use @unstable instead?

that's great idea ! we would need to get this implemented to api-extractor / or eventually introduce custom tooling

Author: https://github.com/microsoft/rushstack/issues/3558

- tooling will trim the declaration from a public API surface.
- can be exposed only from `/unstable` API surface
- tolling will NOT TRIM the declaration from `/unstable` API surface.

**Example:**

- unstable api

```ts
// @filename package-a/src/unstable/index.ts

/**
 *
 * @beta
 */
export function doSomething(): void {}

export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/unstable.d.ts
export function doSomething(): void;
export declare function hello(): void;
```

- stable api

```ts
// @filename package-a/src/index.ts

// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
/**
 *
 * @beta
 */
export function doSomething(): void {}
/**
 *
 * @public
 */
export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/index.d.ts
export declare function hello(): void;
```

#### `@public`

It indicates that an API item has been officially released, and is now part of the supported contract for a package. Semver versioning is used and enforced thus the API signature cannot be changed without a MAJOR version increment.

- tooling will NOT trim the declaration from a public API surface.

> - CANNOT be exposed from `/unstable` API surface

Author: >  it seems that everything is implicitly public unless otherwise annotated.?

yeah, API without annotation is implicitly considered as public ->
![image](https://user-images.githubusercontent.com/1223799/174257442-1f93d7f1-b20c-47c9-b54c-8bf6cd8b45d0.png)

We can with enforce it with tooling, so missing annotation will fail on CI/build

![image](https://user-images.githubusercontent.com/1223799/174257607-68b8458b-66a2-4641-812a-6cf15cfd3b6a.png)

>  Do we need this annotation? When would I use it?

It might feel like unnecessary churn for some devs to add this annotation all the time, on the other hand:
- it will force to provide reasonable amount of documentation to public APIs which is what we should be doing anyways
- consistency of using modifiers

personally I have no strong opinions on this, besides having consistency

> - CANNOT be exposed from `/unstable` API surface

Reviewer 3: IMO we should enforce it via ESLint to be sure that all APIs are properly annotated.

> - CANNOT be exposed from `/unstable` API surface

Reviewer 4: Thanks for the explanation. That really helped.

As you said, it is an extra bit of typing/work but I like the fact that _everything_ would then always be explicit. We uncovered quite a few accidentally public APIs during bug bash so I think it's worth being strict about.

**Example:**

```ts
// @filename package-a/src/index.ts

/**
 *
 * @internal
 */
export function doSomething(): void {}
/**
 *
 * @public
 */
export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/index.d.ts
export declare function hello(): void;
```

### Open questions:

> - should we keep `_unstable` suffix for identifiers ?

Reviewer 2: I think `_unstable` still has a use - for APIs that we want to allow some of our higher value partners to be able to take advantage of knowing full well that they will need to bear the responsibility for maintaining that, otherwise we would say that our composition model is internal 😅.

Author: sounds good, I added more context to Seans question: https://github.com/microsoft/fluentui/pull/23577#discussion_r899904711

> - should we keep `_unstable` suffix for identifiers ?

Reviewer 4: I think we should keep it. Currently our composition API (e.g., `useComponent_unstable`) uses this in otherwise stable components. I could also see us adding other unstable APIs to stable components (e.g., `someProp_unstable`).

Author: using `_unstable` api internally is not an issue. issue is if those apis are exposed as public package surface, which is already happening in our react-components "sub" packages

![image](https://user-images.githubusercontent.com/1223799/174261280-002989ad-def9-443f-8869-0564e967658c.png)

this RFC proposes to expose this packages only from `/unstable` api of particular package, which would be enforced by tooling. This would add us following benefits:
- no churn in package apis 
- stylistic thing: no  `_unstable` churn in identifiers anymore, we can use `_` suffix maybe to simplify (although prefix is a common pattern react hooks are playing against that unfortunately )

Reviewer 4: Let me make sure my understanding is correct:

1. `import { MenuProvider } from '@fluentui/react-menu'`
2. `import { useMenuContext } from '@fluentui/react-menu/unstable'` (currently this is exported as `useMenuContext_unstable` from the stable import path)
3. This would be mirrored in `@fluentui/react-components`

I still think it's nice to have `_unstable` in the name for identifiers like `useMenuContext_unstable`. It's just that much more explicit and it's easy enough to find and replace or codemod when `_unstable` is dropped. The extra explicitness feels nice to me because our consumers should avoid our unstable APIs if at all possible. This issue isn't a blocker for me though.

Author: > Let me make sure my understanding is correct:

yes thats correct 🙌

> I still think it's nice to have _unstable in the name for identifiers like useMenuContext_unstable

yeah sounds good, at least internally it will be more explicit that we are using unstable on our own. but really having a `_` would to the "trick" as well -> `useMenuContext_`

```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/unstable.d.ts
export function doSomething(): void;
export declare function hello(): void;
```

- stable api

```ts
// @filename package-a/src/index.ts

// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
/**
 *
 * @beta
 */
export function doSomething(): void {}
/**
 *
 * @public
 */
export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/index.d.ts
export declare function hello(): void;
```

#### `@public`

It indicates that an API item has been officially released, and is now part of the supported contract for a package. Semver versioning is used and enforced thus the API signature cannot be changed without a MAJOR version increment.

- tooling will NOT trim the declaration from a public API surface.
- CANNOT be exposed from `/unstable` API surface

**Example:**

```ts
// @filename package-a/src/index.ts

/**
 *
 * @internal
 */
export function doSomething(): void {}
/**
 *
 * @public
 */
export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/index.d.ts
export declare function hello(): void;
```

> ### Summary

Author: Please check out the summary with Fufure work paragraph - based on our discussion if it makes sense to you and if we  can proceed.

@layershifter @spmonahan @ling1726 

cheers!

# RFC: Package public API accessors

---

_@hotell_

## Summary

> We wanna control which parts of API surface of our packages are being exposed to public.

Reviewer 5: > 
```suggestion
We want to control which parts of API surface of our packages are being exposed to public.
```

## Background

> During v9 pre-release phase we ran into various situation where we didn't want to propagate APIs as ready to go, thus we established `_unstable` identifiers naming suffix to mark those as unstable. To complement this approach we also introduced `*/unstable` deep imports for API's that are not STABLE. This works to some extent but fails short as its a manual work that is not checked by automation. This led to scenarios that we unintentionally exposed INTERNAL apis to public which led to Breaking changes which we wanna avoid at all costs as we follow proper semver.

Reviewer 5: ```suggestion
During v9 pre-release phase we ran into various situation where we didn't want to propagate APIs as ready to go, thus we established `_unstable` identifiers naming suffix to mark those as unstable. To complement this approach we also introduced `*/unstable` deep imports for API's that are not STABLE. This works to some extent but fails short as it's a manual work that is not checked by automation. This led to scenarios that we unintentionally exposed INTERNAL apis to public which led to breaking changes which we want to avoid at all costs as we follow proper semver.
```

## Problem statement

We expose all APIs of our packages as a PUBLIC from barrel files (`src/index.ts`) to consumers, which introduces problems as we can expose apis that were supposed to be used only internally within our repo or only be available in particular pre-release stage (ALPHA/BETA versions).

## Detailed Design or Proposal

> We wanna provide style guide/common approach how to annotate package APIs with proper ACCESS modifiers which will be processed/verified by tooling automation.

Reviewer 5: ```suggestion
We want to provide style guide/common approach how to annotate package APIs with proper ACCESS modifiers which will be processed/verified by tooling automation.
```

- tolling will NOT TRIM the declaration from `/unstable` API surface.

**Example:**

- unstable api

```ts
// @filename package-a/src/unstable/index.ts

/**
 *
 * @alpha
 */
export function doSomething(): void {}

export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/unstable.d.ts
export function doSomething(): void;
export declare function hello(): void;
```

- stable api

```ts
// @filename package-a/src/index.ts

// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
/**
 *
 * @alpha
 */
export function doSomething(): void {}
/**
 *
 * @public
 */
export function hello(): void {}
```

↓↓ transforms to ↓↓

```ts
// @filename package-a/dist/index.d.ts
export declare function hello(): void;
```

#### `@beta`

It indicates that an API item has been released as a preview or for experimental purposes. Third parties are encouraged to try it and provide feedback. However, a “beta” API should NOT be used in production, because it may be changed or removed in a future version.

- tooling will trim the declaration from a public API surface.
- can be exposed only from `/unstable` API surface

> - tolling will NOT TRIM the declaration from `/unstable` API surface.

Reviewer 5: ```suggestion
- tooling will NOT TRIM the declaration from `/unstable` API surface.
```

export declare function hello(): void;
```

### Summary

Access modifiers introduced in this document will create following package API patterns:

```mermaid
graph TD

> A["@alpha,@beta"] -- access from path --> U["@fluentui/pacakge-name/unstable"]
> B["@public"] -- access from path --> S["@fluentui/pacakge-name"]

Reviewer 3: ```suggestion
A["@alpha,@beta"] -- access from path --> U["@fluentui/package-name/unstable"]
B["@public"] -- access from path --> S["@fluentui/package-name"]
```
